nimare/generate.py

```diff
@@ -0,0 +1,290 @@
+        )
+    else:
+        raise NotImplementedError("Only coordinates for the MNI atlas has been defined")
+
+    # use a template to find all "valid" coordinates
+    template_data = template_img.get_fdata()
+    possible_ijks = np.argwhere(template_data)
+    # number of "convergent" foci each study should report
+    if isinstance(n_foci, int):
+        n_focis = [n_foci] * n_studies
+    else:
+        if len(n_foci) != n_studies:
+            raise ValueError("n_foci must be the same length as n_studies")
+        n_focis = n_foci
+    # foci_coords are to be generated randomly
+    if foci_coords is None:
+        foci_idxs = rng.choice(range(possible_ijks.shape[0]), max(n_focis), replace=False)
+        # if there are no foci_idxs, give a dummy coordinate (0, 0, 0)
+        ground_truth_foci_ijk = (
+            possible_ijks[foci_idxs] if foci_idxs.size else np.array([(0, 0, 0)])
+        )
+        foci_coords_num = ground_truth_foci_ijk.shape[0]
+    # foci_coords were specified by the caller
+    elif isinstance(foci_coords, list):
+        ground_truth_foci_ijk = [mm2vox(coord, template_img.affine) for coord in foci_coords]
+        foci_coords_num = len(ground_truth_foci_ijk)
+    else:
+        raise ValueError("foci_coords must be a list or None")
+
+    if not foci_weights:
+        foci_weights = [1] * foci_coords_num
+    if not isinstance(foci_weights, list):
+        raise ValueError("foci_weights must be a list")
+    elif len(foci_weights) != foci_coords_num:
+        raise ValueError("foci_weights must be the same length as foci")
+
+    if isinstance(fwhm, float):
+        fwhms = [fwhm] * foci_coords_num
+    elif isinstance(fwhm, list):
+        fwhms = [float(f) for f in fwhm]
+        if len(fwhms) != foci_coords_num:
+            raise ValueError(
+                ("fwhm must be a list the same size as the number of foci or a float")
+            )
+    else:
+        try:
+            fwhms = [float(fwhm)] * foci_coords_num
+        except ValueError:
+            raise ValueError("fwhm must be a float or a list")
+
+    foci_dict = {}
+    # generate study specific foci for each ground truth focus
+
+    kernels = [get_ale_kernel(template_img, fwhm)[1] for fwhm in fwhms]
+    weighted_prob_map = sum(
+        [
+            compute_ma(template_data.shape, np.atleast_2d(ground_truth_focus), kernel) * weight
+            for kernel, ground_truth_focus, weight in zip(
+                kernels, ground_truth_foci_ijk, foci_weights
+            )
+        ]
+    )
+    # handle case when sum is 0 by adding a small amount
+    weighted_prob_map = weighted_prob_map / np.sum(weighted_prob_map) + DEF_TINY
+    weighted_prob_map_ijks = np.argwhere(weighted_prob_map)
+    weighted_prob_vector = weighted_prob_map[np.nonzero(weighted_prob_map)]
+
+    if isinstance(n_noise_foci, int):
+        n_noise_foci_list = [n_noise_foci] * n_studies
```

Comment:
I think you can just overwrite `n_noise_foci`, it doesn't seem like it's re-used.

Reply:
good idea, done!